File: DataStructure/Algorithm/Searching/BinarySearch.java

import java.util.*;


class BinarySearchDemo {
    
    static int binarySearch(int arr[], int val,int i,int j) {

        if(i<=j) {  
        int mid= (i+j)/2;
        if(arr[mid]==val)
            return mid;
        else if(arr[mid] < val)
            return binarySearch(arr, val, mid+1,j );
        else if(arr[mid] > val) 
            return binarySearch(arr, val, i, mid-1);
        
        }
        return -1;
    }
}

class Main {

    public static void main(String[] args) {
        
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the size of array: ");
        int n  = sc.nextInt();

        
        int arr[] = new int[n];

        System.out.println("Enter the elements in the array in sorted form: ");

        
        for(int i=0;i<n;i++) {

            arr[i] = sc.nextInt();
        }

        System.out.print("Enter the element to search in the array: ");
        int find = sc.nextInt();

        
        int ind  = BinarySearchDemo.binarySearch(arr, find, 0, n-1);

        if(ind==-1)
            System.out.println("The element "+find+" is not present in the array");   
        else
            System.out.println("The element "+find+" is present at index "+ind);
            
        
    
    }

    
}

File: DataStructure/Core2Web/Searching/BinarySearch.c

//Binary Search

#include <stdio.h>

//Iterative Way
/*int binarySearch(int arr[],int size,int ele) {

	int start = 0;
	int end = size-1;
	int mid;

	while(start<=end) {
		mid = start + (end-start)/2;
		if(arr[mid]==ele)
			return mid;
		if(arr[mid]>ele)
			end = mid-1;
		else
			start= mid+1;;
	}
	return -1;
}
*/


//Using Recursion
int binarySearch(int arr[],int start,int end,int ele) {

	if(start<=end) {

		int mid = start + (end-start)/2;

		if(arr[mid]==ele)
			return mid;
		if(arr[mid]>ele) 
			return binarySearch(arr,start,mid-1,ele);
		return binarySearch(arr,mid+1,end,ele);
	}
	return -1;
}

void main() {

	int arr[] = {3,27,31,125,225,350,729,830,990,1120};

	int size = sizeof(arr)/sizeof(int);
	int start = 0;
	int end = size-1;

	int ele;

	printf("Enter element to search: ");
	scanf("%d",&ele);

	int index = binarySearch(arr,start,end,ele);

	if(index!=-1)
		printf("The element is present at index: %d\n",index);
	else
		printf("The element is not present in the array\n");
}
